### PR TITLE
#112: Fixed Parquet repeated reader bugs

### LINKED ISSUE #112: Parquet repeated converters does not work correctly in nested types
## Situation

Parquet repeated group converters are created once for the whole group. This does not work well when they are nested in another group type.

Example Parquet schema:

```
message test {
  optional group maps (MAP) {
    repeated group key_value {
      required binary name (UTF8);
      repeated group values {
        required int32 year;
        optional int32 height;
        optional int32 weight;
      }
    }
  }
}
```

All the converted values are added to the first key since they are not recreated per each values.

## Acceptance Criteria

- Fixed the bug with repeated converters

### PR DESCRIPTION
**Pull request by morazow**
Fixes #112.

### CODE REVIEW

File: doc/user_guide/user_guide.md [hunk 1]
```diff
@@ -473,6 +473,51 @@ If either of the tags are not set, then it is read as a `null` value.
 | group                | MAP                   | VARCHAR(n)                      |
 | group                | REPEATED              | VARCHAR(n)                      |
 
+#### Parquet Repeated Types
+
+Parquet data type can repeated on a single field or on the group of fields.
+Repeated fields and repeated groups with a single field are imported as JSON
+array string.
+
+For example, the following two schemas are imported as JSON array strings.
```
Comment (jakobbraun) [inline doc/user_guide/user_guide.md:482]: passive

File: doc/user_guide/user_guide.md [hunk 2]
```diff
@@ -473,6 +473,51 @@ If either of the tags are not set, then it is read as a `null` value.
...
+
+```
+message parquet_schema {
+  repeated group person {
+    required binary name (UTF8);
+  }
+}
+```
+
+Both of these Parquet types are imported as a JSON array `["John","Jane"]`.
```
Comment (jakobbraun) [inline doc/user_guide/user_guide.md:498]:
```suggestion
The Parquet importer imports both of these Parquet as a JSON array `["John","Jane"]`.
```

File: doc/user_guide/user_guide.md [hunk 3]
```diff
@@ -473,6 +473,51 @@ If either of the tags are not set, then it is read as a `null` value.
...
+
+On the other hand, a repeated group with multiple fields are imported as JSON
```
Comment (jakobbraun) [inline doc/user_guide/user_guide.md:500]: passive

File: src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala [hunk 4]
```diff
@@ -364,8 +242,20 @@ sealed trait ArrayConverter {
     }
     elementConverter
   }
-  def start(): Unit = dataHolder.reset()
-  def end(): Unit = parentDataHolder.put(index, dataHolder.getValues())
+
+  def start(): Unit = {
+    dataHolder.reset()
+    if (elementConverter.isInstanceOf[RepeatedConverter]) {
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala:248]: How about adding partentStartto the interface, all it every time and only implement it for `RepeatedConverter``.
I've the feeling that it's bad design to have a condition on the type here.

File: src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala [hunk 5]
```diff
@@ -420,10 +310,20 @@ final case class ArrayGroupConverter(
 
   override def createElementConverter(): Converter = new GroupConverter {
     val innerConverter = ParquetConverterFactory(elementType, index, dataHolder)
+
     override def getConverter(index: Int): Converter = innerConverter
-    override def start(): Unit = {}
-    override def end(): Unit = {}
+
+    override def start(): Unit =
+      if (innerConverter.isInstanceOf[RepeatedConverter]) {
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala:317]: same as above

File: src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala [hunk 6]
```diff
@@ -481,13 +381,22 @@ final case class MapConverter(groupType: GroupType, index: Int, parentDataHolder
       } else {
         valuesConverter
       }
-    override def start(): Unit = {}
-    override def end(): Unit = {}
+
+    override def start(): Unit =
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala:385]: same as above

File: src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala [hunk 7]
```diff
@@ -499,8 +408,22 @@ abstract class AbstractStructConverter(
   private[this] val converters = createFieldConverters()
 
   override final def getConverter(fieldIndex: Int): Converter = converters(fieldIndex)
-  override final def start(): Unit = dataHolder.reset()
-  override final def end(): Unit = endOperation()
+  override final def start(): Unit = {
+    dataHolder.reset()
+    for { i <- 0 until size } {
+      if (converters(i).isInstanceOf[RepeatedConverter]) {
+        converters(i).asInstanceOf[RepeatedConverter].parentStart()
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala:415]: same as above

**Review by jakobbraun — CHANGES_REQUESTED** [review]

File: src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala (hunk 4 above)
Comment (morazow) [reply src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala]: Nice catch!
I have updated the converters, override only in repeated ones.

**Review by jakobbraun — APPROVED** [review]

**Pull request merged**